src/Microsoft.FeatureManagement/ConfigurationFeatureSettingsProvider.cs

```diff
@@ -39,17 +40,53 @@ public void Dispose()
             _changeSubscription = null;
         }
 
-        public IFeatureSettings TryGetFeatureSettings(string featureName)
+        public Task<FeatureSettings> GetFeatureSettingsAsync(string featureName)
         {
+            if (featureName == null)
+            {
+                throw new ArgumentNullException(nameof(featureName));
+            }
+
             if (Interlocked.Exchange(ref _stale, 0) != 0)
             {
                 _settings.Clear();
             }
 
-            return _settings.GetOrAdd(featureName, (name) => ReadFeatureSettings(name));
+            //
+            // Query by feature name
+            FeatureSettings settings = _settings.GetOrAdd(featureName, (name) => ReadFeatureSettings(name));
+
+            return Task.FromResult(settings);
         }
 
-        private IFeatureSettings ReadFeatureSettings(string featureName)
+        //
+        // The async key word is necessary for creating IAsyncEnumerable.
+        // The need to disable this warning occurs when implementaing async stream synchronously. 
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        public async IAsyncEnumerable<FeatureSettings> GetAllFeatureSettingsAsync()
+#pragma warning restore CS1998
+        {
+            if (Interlocked.Exchange(ref _stale, 0) != 0)
+            {
+                _settings.Clear();
+            }
+
+            var featureSettings = new List<FeatureSettings>();
+
+            //
+            // Query all
+            foreach (string featureName in GetFeatureConfigurationSections().Select(s => s.Key))
+            {
+                FeatureSettings settings = _settings.GetOrAdd(featureName, (name) => ReadFeatureSettings(name));
```

Comment:
It seems possible the `ReadFeatureSettings` is dealing with a different `IConfigurationSection` than what are returned from `GetFeatureConfigurationSections` in the `foreach` clause when the configuration is changed after the `Interlocked.Exchange` check.

We could add an overload of `ReadFeatureSettings` that takes an optional `IConfigurationSection`.

Reply:
We could, but wouldn't we want to return the latest settings of the feature?

The `GetFeatureConfigurationSections().Select(s => s.Key)` line is to enumerate over all the feature names that are in the system, not necessarily to acquire their current state. We pull the state/settings when the caller moves the enumerator. Maybe I could put a comment for that?

I think the real downside here is that feature names could be added into the IConfiguration system while the caller is enumerating over the returned IAsyncEnumerable and we would miss them. However, the fix for that would involve a lot of calls to retrieve configuration sections and I'm not sure it's worth it.

Reply:
We do want the latest settings. That's why we called `_settings.Clear()` at the beginning of the function. But once we passed that line and made the first call to `GetFeatureConfigurationSections()`, we should stick with that snapshot of configuration. We shouldn't allow configuration changes while we enumerating the collection. This could cause all sorts of hard-to-debug issues when it happens.

I am thinking of an implementation something like this

```cs
foreach (IConfigurationSection featureSection in GetFeatureConfigurationSections())
{
  FeatureSettings settings = _settings.GetOrAdd(featureSection.Key, (featureSection) => ReadFeatureSettings(featureSection));

  ... ...
}

// ... and an overload of ReadFeatureSettings
private FeatureSettings ReadFeatureSettings(IConfigurationSection featureSection)
{
  // This overload will NOT call GetFeatureConfigurationSections(). 
  // It will simply use the IConfigurationSection that is passed in.
  // We can change the current ReadFeatureSettings(string featureName) to call this one.
}

Reply:
Discussed offline. IConfigurationSection does not provide snapshot behavior so there is no way to capture the point-in-time state of all features at the initial invocation of `GetAllFeatureSettingsAsync()`. Thus we will find all feature names that exist at this time, and as the async iterator is moved through we will return the latest state of that feature, or a "default" `FeatureSettings` object if the feature's configuration section was deleted during the async iteration.